data/awsrds/v0/main.go

```diff
@@ -0,0 +1,448 @@
+//go:generate compogen readme ./config ./README.mdx
+package awsrds
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/rds/rdsutils"
+
+	"github.com/instill-ai/component/base"
+	"github.com/jmoiron/sqlx"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	_ "github.com/denisenkom/go-mssqldb"
+	_ "github.com/go-sql-driver/mysql"
+	_ "github.com/jackc/pgx/v5/stdlib"
+	_ "github.com/sijms/go-ora/v2"
+)
+
+const (
+	taskUpsert = "TASK_UPSERT"
+	taskSelect = "TASK_SELECT"
+	taskDelete = "TASK_DELETE"
+	taskQuery  = "TASK_QUERY"
+)
+
+//go:embed config/definition.json
+var definitionJSON []byte
+
+//go:embed config/setup.json
+var setupJSON []byte
+
+//go:embed config/tasks.json
+var tasksJSON []byte
+
+var once sync.Once
+var comp *component
+
+type component struct {
+	base.Component
+}
+
+type execution struct {
+	base.ComponentExecution
+}
+
+func Init(bc base.Component) *component {
+	once.Do(func() {
+		comp = &component{Component: bc}
+		err := comp.LoadDefinition(definitionJSON, setupJSON, tasksJSON, nil)
+		if err != nil {
+			panic(err)
+		}
+	})
+	return comp
+}
+
+func (c *component) CreateExecution(sysVars map[string]any, setup *structpb.Struct, task string) (*base.ExecutionWrapper, error) {
+	return &base.ExecutionWrapper{Execution: &execution{
+		ComponentExecution: base.ComponentExecution{Component: c, SystemVariables: sysVars, Setup: setup, Task: task},
+	}}, nil
+}
+
+type Config struct {
+	DBUser               string
+	DBPassword           string
+	DBName               string
+	DBHost               string
+	DBPort               string
+	Region               string
+	DBAWSAccessKeyID     string
+	DBAWSSecretAccessKey string
+}
+
+func LoadConfig(e *execution) *Config {
+	return &Config{
+		DBUser:               getUser(e.Setup),
+		DBPassword:           getPassword(e.Setup),
+		DBName:               getName(e.Setup),
+		DBHost:               getHost(e.Setup),
+		DBPort:               getPort(e.Setup),
+		Region:               getRegion(e.Setup),
+		DBAWSAccessKeyID:     getAWSAccessKeyID(e.Setup),
+		DBAWSSecretAccessKey: getAWSSecretAccessKey(e.Setup),
+	}
+}
+
+var engines = []string{
+	"postgresql://%s:%s@%s/%s", "sqlserver://%s:%s@%s%s", "oracle://%s:%s@%s/%s", "%s:%s@tcp(%s)/%s",
+}
+
+// newClient creates a new SQL client with traditional username and password
+func newClient(e *execution) (*sqlx.DB, error) {
+	cfg := LoadConfig(e)
+
+	DBEndpoint := fmt.Sprintf("%v:%v", cfg.DBHost, cfg.DBPort)
+
+	// Test every engines to find the correct one
```

Comment:
@chuang8511 

At first, I was thinking of letting the user do so, yet I feel like the component UI is already vertically long enough due to the setup connection. I noticed that every DB being created in AWS RDS has a unique sort of string in its endpoint, making it distinct from one another. The purpose of this design is to:

- Reduce the component UI's height
- Provide convenience for the user since it serves as an auto-detect system

Please let me know your thoughts. I'm personally not sure whether this design will fit into the current development.

Reply:
@zidanehakim 
> Reduce the component UI's height

It is a problem that we should adjust from UI design & frontend in my opinion. 
Some other components also contain this situation.
Thanks for considering that! It is good to have this perspective!

> Provide convenience for the user since it serves as an auto-detect system

auto-detect is good. But, I am not so sure if the users contain 2 sql server with the same setting.
How about we let users choose "postgres", "mysql", "oracle", ..., and we set the raw string for them?
Will it be better?

Reply:
@chuang8511 

If the user has exactly the same setting for 2 different engines, that would be just on master account, port, and region. However, the host would be always different since it has that unique string kind of id.

For example, this is my host
"database.c9y80ugco7tr.ap-southeast-2.rds.amazonaws.com",

The 'c9y80ugco7tr' is the unique string i've mentioned before

Reply:
Ah! I see! Thanks for your explanation!

Another question
How long does each connection take?
`db, err = sqlx.Open("mysql", dsn)`

Will it influence UX eventually?

Reply:
@chuang8511 

After performing end-to-end tests on the component, I found that each task took around 2 seconds at most. I believe the connection is quite fast and won't significantly impact UX. Note that my database is small, with only 2 tables and around 20 entries each, so I dont know if it scales larger.

Reply:
@chuang8511 

I would like to ask for your suggestion. For this component, is it necessary to have the data as an array of JSON inputs to insert multiple entries without using an iterator?

For instance, if I have an array of JSON objects in one line:

css
Copy code
[{"name":"zidane1"},{"name":"zidane2"}]
I can put it in the data field right away as:

css
Copy code
${[{"name":"zidane1"},{"name":"zidane2"}]}
Would this be better? I was thinking that if the user has dozens of data to insert, it would be hard to specify each one individually. Having a one-liner like this would be more convenient.

Reply:
@zidanehakim 
Thanks for testing!

> I found that each task took around 2 seconds at most.

Though the indicator here needs to be carefully defined in the future, we still can easily follow some standards if it does not take too much effort. Process connection for 2 sec is too long in the general standard.

[source of reference](https://blog.sentry.io/whats-the-difference-between-api-latency-and-api-response-time/#:~:text=Just%20to%20have%20a%20number,to%20abandon%20the%20application%2Fwebsite.)
<img width="808" alt="image" src="https://github.com/instill-ai/component/assets/64747455/a7563125-02ad-4245-b039-8de9d3c4fbb0">


>  Note that my database is small, with only 2 tables and around 20 entries each, so I dont know if it scales larger.

Just share, I thought the connection time will be influenced by table & entries count.

Reply:
@zidanehakim 

> I would like to ask for your suggestion. 

I am not quite sure if I understand you correctly. 
Please correct me if you think my response doesn't make sense.

>  I was thinking that if the user has dozens of data to insert

Inserting multiple data in a bulk can reduce the DB connection time, which increase the performance
So, in my imagination, I'd insert data bulk by bulk without using iterator when I build the pipeline.

And, I guess your question is related how a component output referenced by another component input.
I think it is other operator's responsibility, which we (other Instill AI member, probably me or Hui-Tang...) need to take care.

If you think I do not get you, feel free to DM me in Slack. We can quickly huddle to confirm this as well.

Reply:
We should avoid this type of engine detection. Just because `sql.Open` works doesn't mean it will work for all queries. Even though you've listed only four types of engines here, if there are more, some databases might share the same DSN format but have slightly different commands, which could lead to bugs. Let's make the engine selection explicit so users can choose themselves.

Reply:
@donch1989 

Okay, I'll work on adding engine input